fix: plug ingressclass into ctrls

shaneutt opened the pull request:
This PR adds logic to our controllers which configures and handles the `ingress.class` during object reconcilation.

Highlights:
- `Ingress` style objects now require `kubernetes.io/ingress.class: ${CLASS}` in order to be resolved
- Removing the ingress class removes the object from cache/configuration
- Integration tests now prove the functionality above :point_up: 

Resolves https://github.com/Kong/kubernetes-ingress-controller/issues/1220

rainest commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -20,103 +20,114 @@ const outputFile = "controllers/configuration/zz_generated_controllers.go"
+		Plural:                   "kongingresses",
+		URL:                      "configuration.konghq.com",
+		CacheType:                "KongIngress",
+		RequiresIngressClassName: true,
```
Despite the (evergreen source of confusion) name, KongIngress is not an Ingress-like, and is only loaded when some other resource requests it. It does not use class information itself.
```suggestion
		RequiresIngressClassName: false,
```

rainest commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -20,103 +20,114 @@ const outputFile = "controllers/configuration/zz_generated_controllers.go"
+		Plural:                   "kongplugins",
+		URL:                      "configuration.konghq.com",
+		CacheType:                "Plugin",
+		RequiresIngressClassName: true,
```
KongPlugin can only be loaded via other resources (versus KongClusterPlugin, which can be loaded independently if global), so it does not require a class.
```suggestion
		RequiresIngressClassName: false,
```

rainest commented on railgun/manager/manager.go:
```diff
@@ -271,86 +275,99 @@ func Run(ctx context.Context, c *Config) error {
+				Log:              ctrl.Log.WithName("controllers").WithName("KongIngress"),
+				Scheme:           mgr.GetScheme(),
+				Proxy:            prx,
+				IngressClassName: c.IngressClassName,
```
Ditto above:
```suggestion
```

rainest commented on railgun/manager/manager.go:
```diff
@@ -271,86 +275,99 @@ func Run(ctx context.Context, c *Config) error {
+				Log:              ctrl.Log.WithName("controllers").WithName("KongPlugin"),
+				Scheme:           mgr.GetScheme(),
+				Proxy:            prx,
+				IngressClassName: c.IngressClassName,
```
Ditto above:
```suggestion
```

rainest reviewed (dismissed):
The set of resources that require a class annotation in the current proposal doesn't match up with the [1.x set](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/references/annotations/#kubernetesioingressclass). Those are resources which the controller [loads independent of other resources](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/concepts/ingress-classes/#loading-resources-by-class).

Secrets with the `ca-cert` label and KNative Ingresses are omitted, but neither are supported in Railgun yet.

It looks like this only handles the annotation, and not the Ingress V1 `ingressClassName` field; we'll need to handle that also. 1.x's `ingressClassName` handling differs from [the spec](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1453-ingress-api#interoperability-with-previous-annotation) and treats `ingressClassName` identically to the annotation (if the value matches the configured class, load it). We don't do anything with `IngressClass` resources. We do implement the annotation/field preference from the spec, and [use the annotation value if both are present](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/pkg/store/store.go#L188-L196).

shaneutt commented:
> The set of resources that require a class annotation in the current proposal doesn't match up with the [1.x set](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/references/annotations/#kubernetesioingressclass). Those are resources which the controller [loads independent of other resources](https://docs.konghq.com/kubernetes-ingress-controller/1.2.x/concepts/ingress-classes/#loading-resources-by-class).
> 
> Secrets with the `ca-cert` label and KNative Ingresses are omitted, but neither are supported in Railgun yet.
> 
> It looks like this only handles the annotation, and not the Ingress V1 `ingressClassName` field; we'll need to handle that also. 1.x's `ingressClassName` handling differs from [the spec](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1453-ingress-api#interoperability-with-previous-annotation) and treats `ingressClassName` identically to the annotation (if the value matches the configured class, load it). We don't do anything with `IngressClass` resources. We do implement the annotation/field preference from the spec, and [use the annotation value if both are present](https://github.com/Kong/kubernetes-ingress-controller/blob/v1.2.0/pkg/store/store.go#L188-L196).

SGTM, I've added the changes you suggested and added support for IngressClassName spec attr for netv1, lmkwyt :+1:

shaneutt commented:
Also, the tests are getting somewhat terse, I've created a followup to refactor them at a later time: https://github.com/Kong/kubernetes-ingress-controller/issues/1286

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 func (r *{{.PackageAlias}}{{.Type}}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+{{- if .AcceptsIngressClassNameAnnotation}}
+	preds := predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		if v, ok := obj.GetAnnotations()["kubernetes.io/ingress.class"]; ok {
```
the `"kubernetes.io/ingress.class"` magic string already exists as a const: `annotations.IngressClassKey`

shaneutt replied:
acbe0bcaccf4231189496796b60ecdf9cd97f291 resolves this (across the entire code base)

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 // SetupWithManager sets up the controller with the Manager.
 func (r *{{.PackageAlias}}{{.Type}}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+{{- if .AcceptsIngressClassNameAnnotation}}
+	preds := predicate.NewPredicateFuncs(func(obj client.Object) bool {
```
Implementing such (somewhat complex) functions in code generation boilerplate:
- makes the boilerplate harder to read
- makes the boilerplate harder to edit - instead of editing Go code, you edit an opaque huge string in the generator
- makes the logic impossible to unit test

Suggestion: implement and unit test the below function in a non-generated Go package, and then use in a 1-liner fashion in the generated code

`func IngressClassMatcherPredicateFuncs(desiredIngressClass string) predicate.Funcs`

In that case you probably don't need both `AcceptsIngressClassNameAnnotation` and `AcceptsIngressClassNameSpec`. The function above can try match the `ingressClassName` field only on types it expects to have that field (and have a safe failure mode of not matching if the type is not identified as such)

shaneutt replied:
To these points:

> makes the boilerplate harder to read

It's true that if you have no way for your editor to selectively identify and syntax highlight these strings differently that is is harder to read.

> makes the boilerplate harder to edit - instead of editing Go code, you edit an opaque huge string in the generator

Also true to some extent.

> makes the logic impossible to unit test

This part is incorrect, it is entirely possible to unit test template logic. I would be in favor of adding the unit tests rather than tearing out the logic, except I do find the readability and edit-ability points above merited so I'm leaning your way.

---

Given your preference and your points about readability/edit-ability I'm inclined to change this as you suggest, I will have a commit up soon :+1:

shaneutt replied:
Resolved by 20458030831a8fd09d7d59aabe2a2e3aef4e0151

mflendrich replied:
The point has been addressed so the considerations below are purely for academic purposes :mortar_board: 

> This part is incorrect, it is entirely possible to unit test template logic

The template is an opaque string of hundreds of lines and is not code. You can test that the generator generates something that passes unit tests, or you can directly unit test the generated stuff, but in that case you kinda defeat the purpose of generation (by relying on the exact generated content).

shaneutt replied:
> The template is an opaque string of hundreds of lines and is not code

I don't follow this assertion, or understand the intention behind it.

Given the Wordnik dictionary definition of "Source Code" (code is very broad):

```
source code
    n.
    Human-readable instructions in a programming language, to be transformed into machine instructions by a compiler, assembler or other translator, or to be carried out directly by an interpreter.
```

Furthermore the [Golang Template Documentation](https://golang.org/pkg/text/template) explicitly [refers](https://golang.org/pkg/text/template/#hdr-Text_and_spaces) to templates as "source" and "code", examples include:

```
"when the template is executed"
```
```
"However, to aid in formatting template source code"
```
```
"only code specified by the template author should run as a result of injecting the template output into a page"
```
What definition of "code" do you prefer that doesn't match Go templates or what am I missing here to better understand the point?

> You can test that the generator generates something that passes unit tests, or you can directly unit test the generated stuff, but in that case you kinda defeat the purpose of generation (by relying on the exact generated content).

I'm not sure I understand what we're getting at here either, if I use the [Agile Definition Of Unit Tests](https://www.agilealliance.org/glossary/unit-test/):

```
[A unit test] is a short program fragment written and maintained by the developers on the product team, which exercises some narrow part of the product’s source code and checks the results. The outcome of a unit test is binary: either “pass” if the program’s behavior is consistent with the recorded expectations, or “fail” otherwise.
```

The template parser when fed the template code can be fed arguments, and given that input produce multiple different outcomes which could either pass or fail `go test` tests requiring only the local system. Why make a distinction?

mflendrich commented on railgun/test/integration/ingress_test.go:
```diff
@@ -81,6 +82,63 @@ func TestMinimalIngress(t *testing.T) {
 		return false
 	}, ingressWait, waitTick)
 
+	t.Logf("removing the ingress.class annotation \"%s\" from ingress %s", ingressClass, ingress.Name)
```
```suggestion
	t.Logf("removing the ingress.class annotation %q from ingress %s", ingressClass, ingress.Name)
```

mflendrich replied:
this also applies to other instances of this pattern

shaneutt replied:
Good call, I missed this when re-reading my own code :+1: 

resolved by d11036a8b66aa1dec2c3654b09892595fa6cf657

mflendrich commented on railgun/test/integration/ingress_test.go:
```diff
@@ -164,3 +222,146 @@ func TestHTTPSRedirect(t *testing.T) {
+			return false
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode == http.StatusNotFound {
```
I think that we should either make a 404 status code suffice (regardless of body content), or implement the body differ as a helper function - because this minute detail adds a lot of weight to the code of the test case.

shaneutt replied:
resolved by 4e9927f59ad2f4d4f09c358b70aea4d864a253b8 after some talk during our sync today :+1:

mflendrich commented on railgun/test/integration/ingress_test.go:
```diff
@@ -81,6 +82,63 @@ func TestMinimalIngress(t *testing.T) {
 	}, ingressWait, waitTick)
 
+	t.Logf("removing the ingress.class annotation \"%s\" from ingress %s", ingressClass, ingress.Name)
+	ingress, err = cluster.Client().NetworkingV1().Ingresses("default").Get(ctx, ingress.Name, metav1.GetOptions{})
```
nit: put the namespace name in a `const`

shaneutt replied:
I forgot about this you're right, changed to `corev1.NamespaceDefault` in 5b94b7e2a3d03e3957078703efa056967aef9640 :+1:

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 type {{.PackageAlias}}{{.Type}}Reconciler struct {
 	client.Client
 
+{{- if or .AcceptsIngressClassNameSpec .AcceptsIngressClassNameAnnotation}}
```
If you put a blank line between `IngressClassName` and other fields, you won't need to be concerned with nice block formatting of these other fields, and the generated code will still pass `gofmt`.

shaneutt replied:
:+1:  1d99e3614a8fb3eb836028ccb88768ce1ae239e2

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *{{.PackageAlias}}{{.Type}}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+{{- if .AcceptsIngressClassNameAnnotation}}
```
suggestion: make predicate funcs injectable into a controller, instead of branching out the code generator.

mflendrich commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -59,3 +60,24 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
 
 	return true, nil
 }
+
```
Whatever implementation choices get made in this PR, please add a comment to all functions dealing with ingress classes in the `annotations` package that they're deprecated in favor of &lt;specific replacement function name&gt;.

Or reuse them.

shaneutt replied:
:+1: 3dbb54261963e216f1fa391daa15555d35301b2b

mflendrich commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -59,3 +60,24 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
+// HasAnnotation is a helper function to determine whether an object has a given annotation, and whether it's
+// to the value provided.
+func HasAnnotation(obj client.Object, key, val string) bool {
+	if v, ok := obj.GetAnnotations()[key]; ok {
+		if v == val {
+			return true
+		}
+	}
+	return false
```
Branchless code is more readable and performant.

```suggestion
	v, ok := obj.GetAnnotations()[key]
	return ok && v == v.val
```

shaneutt replied:
So I got a little curious (just for the sake of curiosity) about whether or not the new solution was actually technically "more performant". I decided nerd out a little bit and take a couple minutes to break it down. I created the following program to represent the performance of the two different control flow mechanisms:

```golang
package main

var z bool
var x bool

func main() {}

func Batman() bool {
	if z {
		if x {
			return true
		}
	}
	return false
}

func Robin() bool {
	return z && x
}
```

And then I compiled it and determined the assembly code that would be produced by the Go compiler (`go1.16.3 linux/amd64`):

```shell
$ go build -buildmode=plugin
$ objdump -Mintel -d test.so > disassembled
```

This is the resulting (Intel syntax) Assembly code:

```asm
00000000000a3e80 <test.Batman>:
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
   a3ea1:	c6 44 24 08 01       	mov    BYTE PTR [rsp+0x8],0x1
   a3ea6:	c3                   	ret

00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
   a3edd:	31 c0                	xor    eax,eax
   a3ee0:	eb f6                	jmp    a3ed8 <test.Robin+0x18>
```

There are four cases to represent given the above assembly control flow:

1. z == false && x == true
2. z == false && x == false
3. z == true && x == false
4. z == true && x == true

The `Batman` function covers both `#1` and `#2` with the following code path running only 5 total instructions:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
```

Then `Batman` covers `#3` with:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3e9a:	c6 44 24 08 00       	mov    BYTE PTR [rsp+0x8],0x0
   a3ea0:	c3                   	ret
```

And finally `#4` with:

```asm
   a3e80:	4c 8b 3d b1 fe 05 00 	mov    r15,QWORD PTR [rip+0x5feb1]        # 103d38 <test.z@@Base-0x305b7>
   a3e87:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e8b:	74 0d                	je     a3e9a <test.Batman+0x1a>
   a3e8d:	4c 8b 3d 94 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe94]        # 103d28 <test.x@@Base-0x305c6>
   a3e94:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3e98:	75 07                	jne    a3ea1 <test.Batman+0x21>
   a3ea1:	c6 44 24 08 01       	mov    BYTE PTR [rsp+0x8],0x1
   a3ea6:	c3                   	ret
```

Which is effectively the same as the solution for `#3`.

As for the `Robin` function when processing `#1` it carries out 7 total instructions:

```asm
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

The above implementation has to execute 4 `mov` instructions (including one zero-extending `movzx`) over the `Batman` function's solution which solves it with 2 `mov` instructions.


As for `#2` with `Robin`:

```asm
00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3edd:	31 c0                	xor    eax,eax
   a3ee0:	eb f6                	jmp    a3ed8 <test.Robin+0x18>
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

Similar to above 7 total instructions are needed (versus the 5 needed by the `Batman` function) including more writing to registers (using `xor`) to do the same job, and additional `jmp` instructions.

Problem `#3` is solved this way:

```asm
00000000000a3ec0 <test.Robin>:
   a3ec0:	4c 8b 3d 71 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe71]        # 103d38 <test.z@@Base-0x305b7>
   a3ec7:	41 80 3f 00          	cmp    BYTE PTR [r15],0x0
   a3ecb:	74 10                	je     a3edd <test.Robin+0x1d>
   a3ecd:	4c 8b 3d 54 fe 05 00 	mov    r15,QWORD PTR [rip+0x5fe54]        # 103d28 <test.x@@Base-0x305c6>
   a3ed4:	41 0f b6 07          	movzx  eax,BYTE PTR [r15]
   a3ed8:	88 44 24 08          	mov    BYTE PTR [rsp+0x8],al
   a3edc:	c3                   	ret
```

And has to do an additional `movzx` over the `Batman` implementation, however it does reduce the number of comparison instructions by one. The solution for problem `#4` is effectively the same as `#3`, but again with the extra `movxz`.

Given all the above if we assume all input options are equal this means that the original solution requires two less instructions to solve the same problem and significantly less moving of bits between registers to do so as well. `Robin` technically has the advantage of using less memory, as it only requires 9 instructions to be loaded vs. 10. Overall this seems to indicate to me that **the original solution was ultimately "more performant"** from a runtime perspective _however there are more factors that could be taken into account_ such as seeing how branch prediction behavior would affect things, e.t.c.

I dug into this for fun (yes, I indeed find this kind of stuff fun :joy:) and because it doesn't cost much time to just compile, dump the binary and look at the resulting solutions in assembly, but I don't want to go further with this as I'm already swimming pretty deep in the waters of the pedantic ocean. It's pretty obvious when we're talking about a few bits in memory and a few extra instructions per call that we're going off the rails a bit as this is not 1981 and the systems which this software run have multiple millions and billions of processor cycles and memory availability that make this kind of thing beyond negligible, and I'm just fine with the second option, but I thought you might find it kinda interesting.

This is also just a shameless plug for my _secret hidden agenda_ that we *rewrite Kubernetes in assembly*!!! :scream:

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -234,14 +267,66 @@ var controllerTemplate = `
 }
 
 // SetupWithManager sets up the controller with the Manager.
 func (r *{{.PackageAlias}}{{.Type}}Reconciler) SetupWithManager(mgr ctrl.Manager) error {
```
IIUC, this implementation seems to ignore the `--process-classless-XXX` flags implemented in https://github.com/Kong/kubernetes-ingress-controller/pull/1262 following the discussion in https://github.com/Kong/kubernetes-ingress-controller/pull/1262#discussion_r622976024 which renders these settings ineffective.

We need to do one of the below:
- implement classless handling throughout Railgun (and integration test it)
- decide to drop classless resources altogether

Following the discussion cited above, 2.0 seems to me like a good moment to do the latter.

rainest replied:
So apparently answer to

> Does continued maintenance impose a burden beyond addition of the flags?

is yes? Arguments in favor of dropping it are then:

- The classless behavior both conflicted with other controllers and was confusing even without them (classless handled, but only if your ingress class is set to "kong"), which is why we made it not the default.
- The new default behavior was added [~8 months ago](https://github.com/Kong/kubernetes-ingress-controller/blob/main/CHANGELOG.md#0100---20200915), which is IMO a reasonably long period to be aware of the new, preferred default.
- The old classless behavior is incompatible with the Ingress V1 [default class spec](https://kubernetes.io/docs/concepts/services-networking/ingress/#default-ingress-class), which labels a single IngressClass the cluster default and assigns all Ingress resources without class information that class.
- It's more complexity to maintain, and that complexity will become worse if we try to somehow support both it and the proper V1 behavior in code (as mutually exclusive settings).

shaneutt replied:
I have created a follow-up item for this issue: https://github.com/Kong/kubernetes-ingress-controller/issues/1293 :bow:

mflendrich replied:
I have two arguments against doing this in a follow-up issue.

- From the eng perspective: because of the classless handling mismatch, this PR adds ingress class functionality that is effectively broken and violates the contract (to the extent of interpretation of classlessness). I am not comfortable with merging in a PR with a well known contract breach for several reasons: bus factor, potential change of priorities ahead, us forgetting to make a fix or making an ineffective one, etc.
- From the PM perspective: repeated behavior of closing an issue with an incomplete solution and creation of a follow-up issue to "fill in the gap" leads to a project where it's impossible to track progress, because we're moving the goalposts with each such created issue and end up "crawling" first over the initial goals, then over fixes for the gaps, then for fixes for their fixes, etc.

mflendrich commented on railgun/hack/generators/controllers/networking/main.go:
```diff
@@ -20,103 +20,125 @@ const outputFile = "controllers/configuration/zz_generated_controllers.go"
-		CacheType:          "Endpoint",
+		PackageImportAlias:                "corev1",
+		PackageAlias:                      "CoreV1",
+		Package:                           "k8s.io/api/core/v1",
```
nit: `const` out the `Package`s

shaneutt replied:
:+1: e229d5a4440941faf39f7eee6d95ea47aab2a8e1 resolves

mflendrich commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -59,3 +67,79 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
+
+// HasAnnotation is a helper function to determine whether an object has a given annotation, and whether it's
+// to the value provided.
+func HasAnnotation(obj client.Object, key, expectedValue string) bool {
```
I think that most of these functions added in this file (except for the predicate funcs generator) belong to `pkg/annotations`.

shaneutt replied:
I tend to disagree because these functions are commonly operating on `controller-runtime`'s `client.Object`, I don't think introducing that in `pkg/annotations` which previously had no awareness of CR, and which is technically a part of our public API (this package is intentionally under `internal/` to keep us free to move things around later when the dust settles) is something I would like to do within the scope of this PR, I'm assuming this is not a blocker, but if you feel there's more that needs to be said about this please let me know. :bow:

mflendrich commented on railgun/internal/ctrlutils/utils.go:
```diff
@@ -59,3 +67,79 @@ func IsAPIAvailable(mgr ctrl.Manager, obj client.Object) (bool, error) {
+
+// IsIngressClassAnnotationConfigured determines whether an object has IngressClassName field in its spec and whether the value
+// matches the provide IngressClassName (and is therefore an object configured to be reconciled by that class).
+func IsIngressClassSpecConfigured(obj client.Object, expectedIngressClassName string) bool {
```
1. Consider using a simple type switch instead of reflection, for two reasons.
    - brevity: less code
    - performance: this is much faster than reflection
    - defensive programming: the current code has a nasty side effect that, when in 2 years from now, us or some other engineer (with no memory of this reflection trick) adds a new type to KIC, it will automatically interpret this field using its name by convention.
```go
switch obj := obj.(type) {
case netv1.Ingress:
  return obj.Spec.IngressClassName != nil && *obj.Spec.IngressClassName == expectedIngressClassName
}

return false
```

shaneutt replied:
Yep, this is a good call. `reflect` is fun, but costly. Changed in rebase :+1:

mflendrich reviewed (commented):
Once we sort out the logging problem, this is good to go :+1:

shaneutt commented on railgun/manager/manager.go:
```diff
@@ -225,9 +223,13 @@ func Run(ctx context.Context, c *Config) error {
+	//        See: https://github.com/bombsimon/logrusr/issues/2
+	//
+	// var logger logr.Logger = logrusr.NewLogger(deprecatedLogger)
+	// ctrl.SetLogger(logger)
```
yes, this was just temporary for testing purposes, it's been dropped out in a rebase against `next` with the patched release :+1:
